linter.py

```diff
@@ -1,57 +1,72 @@
+import logging
+import os
+from SublimeLinter.lint import Linter, util
 
-from SublimeLinter.lint import Linter, util, persist
+
+logger = logging.getLogger('SublimeLinter.plugin.eslint')
 
 
 class Codeclimate(Linter):
     """Provides an interface to codeclimate."""
-
-    syntax = (
-        'ruby on rails',
-        'ruby',
-        'javascript',
-        'python',
-        'go',
-        'html',
-        'css',
-        'markdown',
-        'php',
-        'haskall',
-        'java',
-        'shellscript'
-    )
-    executable = "codeclimate"
+    defaults = {
+        'selector': (
+            'source.css, '
+            'source.go, '
+            'source.haskell, '
+            'source.java, '
+            'source.javascript, '
+            'source.js, '
+            'source.php, '
+            'source.python, '
+            'source.ruby, '
+            'source.scss, '
+            'source.shell, '
+            'text.html'
+        )
+    }
     regex = r'^(?P<line>\d+)(?:-\d+)?:\s(?P<message>.+)$'
     multiline = False
     line_col_base = (1, 1)
     tempfile_suffix = '-'
     error_stream = util.STREAM_BOTH
-    selectors = {}
     word_re = None
-    defaults = {
-        'chdir': "${project}"
-    }
-    inline_settings = None
-    inline_overrides = None
-    comment_re = None
 
     def cmd(self):
-        """Construct a cmd to provide a relative path to 'codeclimate analyze'."""
-        result = [self.executable_path, 'analyze']
-        relative_file_name = None
-        for folder in self.view.window().folders():
-            if self.filename.find(folder) == 0:
-                relative_file_name = self.filename.replace(folder + '/', '')
-        result.append(relative_file_name)
-        persist.debug(result)
-        return result
+        """Set command and run 'codeclimate analyze'."""
+        abs_path = self.filename
+        working_dir = self.get_working_dir()
+
+        msg = 'You try to lint the file %s from outside of SublimeText\'s ' \
+              'working directory (%s), which is not supported by this ' \
+              'SublimeLinter plugin at this time. You can do either:\n' \
+              '* You open the directory containing the current file in' \
+              'a new window. You may have to add a customized ' \
+              '`.codeclimate.yml` configuration to that directory ' \
+              'as well;\n' \
+              '* You change the working directory of the linter plugin in ' \
+              ' the global or the Project\'s settings;\n' \
+              '* You disable the linter plugin in the global or the ' \
+              'Project\'s settings.\n\n' \
+              'Please visit https://github.com/codeclimate/' \
+              'SublimeLinter-contrib-codeclimate for examples.' \
+              '\n' % (abs_path, working_dir)
+
+        if not (abs_path.startswith(os.path.abspath(working_dir) + os.sep)):
+            logger.info(msg)
+            return None
```

Comment:
Instead of `return None` which is not valid, `raise PermanentError("file not part of the working dir")` (error message up to you).  At the top add `from SublimeLinter.lint import PermanentError`. 

You probably want to call `self.notify_unassign()` before the `raise` to not show any errors to the user, basically to unassign the linter from the view.  (Just as `logger.info` does only print in debug mode.)  If you want to indicate an error to the user use `self.notify_failure()` instead and `logger.warn`.  (I don't think that's a use case here because the user cannot fix this, we just ignore arbitrary views from outside the main folder.)

Reply:
I wonder if it's enough to set 

```
defaults = {
	...
	"excludes": ["!${folder::}*"]
}
```

t.i. exclude any filenames not starting with `${folder}`.  And since folder can be empty, with a fallback `${folder::}` since no path can start with a `:`.

Reply:
Ah, maybe not that good since a user can change this.

Reply:
I added the exclusion statement, and it works as predicted. If I have opened a directory and then open a single file from another directory, the linter will not run:

```text
SublimeLinter: linter.py:1042         codeclimate skipped '/Users/run/Desktop/ExternalFormatter.php', excluded by '!/Users/run/Desktop/TEST*'
INFO:SublimeLinter.lint.linter:codeclimate skipped '/Users/run/Desktop/ExternalFormatter.php', excluded by '!/Users/run/Desktop/TEST*'
```

[I also updated][1] my exception handling to:

```python
msg = 'The file \'%s\' is not part of the working dir (%s). ' \
      'Please see the Linter\'s README.md to get further ' \
      'instructions.' % (abs_path, working_dir)

if not (abs_path.startswith(os.path.abspath(working_dir) + os.sep)):
    self.notify_unassign()
    raise PermanentError(msg)
```

I know it is a bit redundant along with the exclusion statement, but I suggest keeping it in place just in case…

[1]:https://github.com/codeclimate/SublimeLinter-contrib-codeclimate/pull/3/files#diff-e7c2c765d5c484aae3f449e93444a2659ddc4bac3b83d10c6ee5c36f6c9e8eebR50

Reply:
Ah, you need to log the message using the `logger` as we don't log the error message, sorry that was probably misleading.   (In `PermanentError(msg)` the `msg` is not for the user; you *could* not set a msg as well.)

I don't think the `excludes` trick is worth it; as the user can set the `working_dir` manually, in that case our `excludes` (`$folder`) would not work or still point to a different directory.

Reply:
I updated the exception handling in doing a `logger.warning` first and then and then unassign and raise the permanent error in https://github.com/codeclimate/SublimeLinter-contrib-codeclimate/pull/3/commits/281e165fbf5bc070b78b5f184b4509694ebd2018.

Ah yes, you're right. I removed the exclusion in https://github.com/codeclimate/SublimeLinter-contrib-codeclimate/pull/3/commits/c44c78ad3c903e15273aa7ae45bb9c15488c2929.